src/Rings/AbelianClosure.jl

```diff
@@ -0,0 +1,781 @@
+  return QabElem(divexact(a.data, b.data), a.c)
+end
+
+function inv(a::QabElem)
+  return QabElem(inv(data(a)), a.c)
+end
+
+################################################################################
+#
+#  Unsafe operations
+#
+################################################################################
+
+function addeq!(c::QabElem, a::QabElem)
+	_c, _a = make_compatible(c, a)
+	addeq!(_c.data, _a.data)
+  return _c
+  #c.data = _c.data
+  #c.c = _c.c
+	#return c
+end
+
+function neg!(a::QabElem)
+	mul!(a.data,a.data,-1)
+	return a
+end
+
+function mul!(c::QabElem, a::QabElem, b::QabElem)
+	a, b = make_compatible(a, b)
+	b, c = make_compatible(b, c)
+  a, b = make_compatible(a, b)
+	mul!(c.data, a.data, b.data)
+	return c
+end
+
+################################################################################
+#
+#  Ad hoc binary operations
+#
+################################################################################
+
+*(a::fmpz, b::QabElem) = QabElem(b.data*a, b.c)
+
+*(a::fmpq, b::QabElem) = QabElem(b.data*a, b.c)
+
+*(a::Integer, b::QabElem) = QabElem(data(b) * a, b.c)
+
+*(a::Rational, b::QabElem) = QabElem(data(b) * a, b.c)
+
+*(a::QabElem, b::fmpz) = b*a
+
+*(a::QabElem, b::fmpq) = b*a
+
+*(a::QabElem, b::Integer) = b*a
+
+*(a::QabElem, b::Rational) = b*a
+
++(a::fmpz, b::QabElem) = QabElem(b.data + a, b.c)
+
++(a::fmpq, b::QabElem) = QabElem(b.data + a, b.c)
+
++(a::Integer, b::QabElem) = QabElem(data(b) + a, b.c)
+
++(a::Rational, b::QabElem) = QabElem(data(b) + a, b.c)
+
++(a::QabElem, b::fmpz) = b + a
+
++(a::QabElem, b::fmpq) = b + a
+
++(a::QabElem, b::Integer) = b + a
+
++(a::QabElem, b::Rational) = b + a
+
+-(a::fmpz, b::QabElem) = QabElem(-(a, data(b)), b.c)
+
+-(a::fmpq, b::QabElem) = QabElem(-(a, data(b)), b.c)
+
+-(a::Integer, b::QabElem) = QabElem(-(a, data(b)), b.c)
+
+-(a::Rational, b::QabElem) = QabElem(-(a, data(b)), b.c)
+
+-(a::QabElem, b::fmpz) = QabElem(-(data(a), b), a.c)
+
+-(a::QabElem, b::fmpq) = QabElem(-(data(a), b), a.c)
+
+-(a::QabElem, b::Integer) = QabElem(-(data(a), b), a.c)
+
+-(a::QabElem, b::Rational) = QabElem(-(data(a), b), a.c)
+
+//(a::QabElem, b::fmpz) = QabElem(data(a)//b, a.c)
+
+//(a::QabElem, b::fmpq) = QabElem(data(a)//b, a.c)
+
+//(a::QabElem, b::Integer) = QabElem(data(a)//b, a.c)
+
+//(a::QabElem, b::Rational) = QabElem(data(a)//b, a.c)
+
+################################################################################
+#
+#  Comparison
+#
+################################################################################
+
+function ==(a::QabElem, b::QabElem)
+  a, b = make_compatible(a, b)
+  return a.data == b.data
+end
+
+function ==(a::QabElem, b::Union{fmpz, fmpq, Integer, Rational})
+  return data(a) == b
+end
+
+function ==(a::Union{fmpz, fmpq, Integer, Rational}, b::QabElem)
+  return b == a
+end
+
+function (b::QabField)(a::Singular.n_unknown{QabElem}) 
+  Singular.libSingular.julia(Singular.libSingular.cast_number_to_void(a.ptr))
+end
```

Comment:
```suggestion
```
Delete this and add an assertion `::elem_type(b)` on the code in mpoly.jl:
```julia
#??? needs to coerce into b? assert parent?
function (b::AbstractAlgebra.Ring)(a::Singular.n_unknown)
  Singular.libSingular.julia(Singular.libSingular.cast_number_to_void(a.ptr))::elem_type(b)
end
```

Reply:
everything passes with the assertion, or (your call) you can do the thing which is probably safer:
```julia
#??? needs to coerce into b? assert parent?
function (b::AbstractAlgebra.Ring)(a::Singular.n_unknown)
  b(Singular.libSingular.julia(Singular.libSingular.cast_number_to_void(a.ptr)))
end
```

Reply:
I do not require any of these changes anymore, so please apply them at your own disgression.